File: LeetCode/Study Plans/30 Days of LC JavaScript Challenge/Day-5.js
// Problem :- 2634. Filter Elements from Array
// Link :- https://leetcode.com/problems/filter-elements-from-array/?gio_link_id=a9a5VZr9

// Solution:
/**
 * @param {number[]} arr
 * @param {Function} fn
 * @return {number[]}
 */
var filter = function(arr, fn) {
    const res = [];
    let j = 0;
    for(let i = 0; i < arr.length; i++) if(fn(arr[i], i)) res[j++] = arr[i];
    return res;
};

File: LeetCode/POTD/2025/January/Day-5.cpp
// Problem Link: https://leetcode.com/problems/shifting-letters-ii/

// Solution:
class Solution {
public:
    string shiftingLetters(string s, vector<vector<int>>& shifts) {
        vector<int> v(s.length(), 0);
        for(auto &shift: shifts) {
            if(shift[2]) {
                v[shift[0]]++;
                if(shift[1]+1 < s.length()) v[shift[1]+1]--;
            } else {
                v[shift[0]]--;
                if(shift[1]+1 < s.length()) v[shift[1]+1]++;
            }
        }
        int numShifts = 0;
        for(int i = 0; i < s.length(); i++) {
            numShifts = (numShifts + v[i]) % 26;
            if(numShifts < 0) numShifts += 26;
            s[i] = 'a' + (s[i] - 'a' + numShifts)%26;
        }
        return s;
    }
};